### PR TITLE
New search page layout, search bug fixes, image fixes

### PR DESCRIPTION
**Pull request by benvogler**
This PR was originally supposed to add filters (which is why there is now a dummy "Within X Miles" filter at the top of the search page) but turned into fixing lots of minor bugs, so I decided to PR this separately before I dive into filter functionality

1. New search results page layout
2. Refactor the Search page into more child components
3. Center the search results map on the location if one is specified (this currently just always uses a zoom of 9 - we will want to adjust this based on the distance filter. We may also want to think about locations such as "United States" that google can give us, which will actually not work well regardless due to the distance filtering)
4. Fix the google api key placement in index.html - it had extra curly braces that was breaking it
5. Show images in search results
6. Fix some spacing issues with the monument page gallery, and don't show the button at the bottom if there's only one image
7. Add clear buttons to the search bars so that you can easily remove your search criteria if you desire (note that this will instantly initiate a search - I think this is the best behavior but it's up for debate)
8. Fix searching for both location and text, as this was previously not using an `AND` in the generated SQL

Here's an example of the new search page, searching for both text and location, and zoomed in on the selected location

### CODE REVIEW

File: client/src/components/Header/SearchBar/LocationSearch/LocationSearch.scss [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    }
+}
+
+.autocomplete-dropdown-container {
+    position: absolute;
+    top: 30px;
+    background-color: #ffffff;
+    border: 1px solid rgba(0,0,0,0.2);
+    // This has to be absurdly large due to the leaflet map's absurdly large z-index
+    z-index: 1000;
```
Comment (njd9419) [inline client/src/components/Header/SearchBar/LocationSearch/LocationSearch.scss]: Nice

File: client/src/components/Header/SearchBar/SearchBar.js [hunk 2]
```diff
@@ -12,20 +12,27 @@ export default class SearchBar extends React.Component {
...
 
     handleTextSearchChange(textSearchQuery) {
         this.setState({textSearchQuery: textSearchQuery});
     }
 
-    handleLocationSearchSelect(lat, lon, address) {
-        this.setState({locationLat: lat, locationLon: lon, locationAddress: address});
+    async handleTextSearchClear() {
+        await this.setState({textSearchQuery: ''});
+        console.log('clear');
```
Comment (njd9419) [inline client/src/components/Header/SearchBar/SearchBar.js]: Debug print?

Comment (benvogler) [reply client/src/components/Header/SearchBar/SearchBar.js]: come on, I would _never_

File: client/src/components/Header/SearchBar/LocationSearch/LocationSearch.scss (hunk 1 above)
Comment (benvogler) [reply client/src/components/Header/SearchBar/LocationSearch/LocationSearch.scss]: I could lower the map's z-index but there might be a reason it's that high 🤷‍♂

File: client/src/components/Header/SearchBar/SearchBar.js [hunk 3]
```diff
@@ -45,17 +51,19 @@ export default class SearchBar extends React.Component {
             d: 25,
             address: locationAddress
         });
-        window.location.replace(`/search/?${queryString}`);
+        window.location.href = `/search/?${queryString}`;
```
Comment (njd9419) [inline client/src/components/Header/SearchBar/SearchBar.js]: Is there a reason this was changed?

File: client/src/components/Header/SearchBar/TextSearch/TextSearch.js [hunk 4]
```diff
@@ -1,4 +1,5 @@
 import React from 'react';
+import PlacesAutocomplete from 'react-places-autocomplete';
```
Comment (njd9419) [inline client/src/components/Header/SearchBar/TextSearch/TextSearch.js]: Unused import?

File: client/src/components/Pagination/Pagination.js [hunk 5]
```diff
@@ -21,8 +21,10 @@ export default class Pagination extends React.Component {
 
     componentDidUpdate(prevProps, prevState, snapshot) {
         const { count, page } = this.props;
-        if (count !== prevProps.count) this.setState({count});
-        if (page !== prevProps.page) {
+        if (count !== prevProps.count && !isNaN(count)) {
```
Comment (njd9419) [inline client/src/components/Pagination/Pagination.js]: Were you running into NaNs?

File: client/src/components/Header/SearchBar/SearchBar.js (hunk 2 above)
Comment (njd9419) [reply client/src/components/Header/SearchBar/SearchBar.js]: Please remove, it's having a huge impact on performance

File: src/main/java/com/monumental/services/MonumentService.java [hunk 6]
```diff
@@ -107,15 +104,29 @@ private void buildDWithinQuery(CriteriaBuilder builder, CriteriaQuery query, Roo
...
-                                  String latitude, String longitude, String distance, Boolean orderByResults) {
+                                  Double latitude, Double longitude, Integer distance, Boolean orderByResults) {
         // TODO: Query for tags
         // TODO: Filters
-        if (searchQuery != null) {
-            this.buildSimilarityQuery(builder, query, root, searchQuery, 0.1, orderByResults);
+
+        List<Predicate> predicates = new ArrayList<>();
+
+        if (searchQuery != null && !searchQuery.trim().equals("")) {
```
Comment (njd9419) [inline src/main/java/com/monumental/services/MonumentService.java]: You should be able to use `!StringHelper.isNullOrEmpty(searchQuery)`

File: src/main/java/com/monumental/services/MonumentService.java [hunk 7]
```diff
@@ -185,15 +198,29 @@ public void getRelatedTags(List<Monument> monuments) {
...
+
+        String capitalizedFieldName = fieldName.substring(0, 1).toUpperCase() + fieldName.substring(1);
 
-        Map<Integer, List<Tag>> tagsMap = new HashMap<>();
-        for (Monument monument : monumentsWithTags) {
-            tagsMap.put(monument.getId(), monument.getTags());
+        Map<Integer, List> map = new HashMap<>();
+        for (Monument monument : monumentsWithRecords) {
+            try {
+                map.put(monument.getId(), (List) Monument.class.getDeclaredMethod("get" + capitalizedFieldName).invoke(monument));
```
Comment (njd9419) [inline src/main/java/com/monumental/services/MonumentService.java]: woof

File: client/src/components/Header/SearchBar/SearchBar.js (hunk 3 above)
Comment (benvogler) [reply client/src/components/Header/SearchBar/SearchBar.js]: `replace` doesn't add a page to the history, so pressing back didn't work right

File: client/src/components/Pagination/Pagination.js (hunk 5 above)
Comment (benvogler) [reply client/src/components/Pagination/Pagination.js]: Yes, something to do with the values defaulting to strings I think...

File: src/main/java/com/monumental/services/MonumentService.java (hunk 7 above)
Comment (benvogler) [reply src/main/java/com/monumental/services/MonumentService.java]: I've gotten _so_ used to be able to do `monument[fieldName]` in Javascript, it pains me to write this kind of reflection

File: src/main/java/com/monumental/services/MonumentService.java (hunk 6 above)
Comment (benvogler) [reply src/main/java/com/monumental/services/MonumentService.java]: Done

File: client/src/components/Header/SearchBar/TextSearch/TextSearch.js (hunk 4 above)
Comment (benvogler) [reply client/src/components/Header/SearchBar/TextSearch/TextSearch.js]: Fixed

**Comment by benvogler** [comment]
I added a new feature to this PR - the map will now automatically fit all markers into it and zoom as close as it can to them (leaving some room to have padding around them so they're not on the edges of the map exactly)

For example, notice the second map is zoomed in more

**Pull request merged**